proposals/bindless.md

```diff
@@ -333,7 +333,80 @@ In implementations this can be done with almost no additional overhead to the cu
+
+```webidl
+partial interface GPUBindGroup {
+    void update(GPUIndex32 binding, GPUBindingResource resource);
+    GPUIndex32 insertBinding(GPUBindingResource resource);
+    void removeBinding(GPUIndex32 binding);
+};
+```
+
+Two ways to update the dynamic binding array are exposed to allow both implicit and explicit allocation of resources in the dynamic binding array.
+`insertBinding` is simpler to use because it defers to the browser's tracking of which slots may be in use and returns to the user the `binding` it placed the `resource` on (or an exception on failure).
+On the other hand `update` gives the application control of where it places binding, which may be useful to allocate contiguous ranges, for hardcoded slots, or when porting code manually allocating in D3D12/Metal/Vulkan already.
+
+Additional internal state is added to `GPUBindGroup`:
+
+ - An `Array<boolean>` called `[[slot_maybe_used]]` of size `this.[[desc]].dynamicArraySize` initially filled with `false` values.
+
+Steps for `GPUBindGroup.update(binding, resource)`:
+
+ - If `this.[[desc]].layout.[[desc]].dynamicArray` is `undefined`, throw an `OperationError`.
+ - Let `slot` be `binding - this.[[desc]].layout.[[desc]].dynamicArray.start`.
+ - If any of the following is not satisfied, throw an `OperationException`:
+
+    - `slot < this.[[desc]].dynamicArraySize`
+    - `!this.[[slot_maybe_used]][slot]`
+
+ - Set `this.[[slot_maybe_used]][slot]`
+ - On the device timeline:
+
+   - If any of the following is not satified, generate a validation error and return.
+
+      - `this` is valid and `destroy()` hasn't been called on it.
+      - If `resource` is not compatible with `this.[[desc]].layout.[[desc]].dynamicArray.type` (TODO [#5374](https://github.com/gpuweb/gpuweb/issues/5374), determine the compatibility rules).
+
+   - Set the entry at `binding` in the bindgroup to `resource`.
+
+Steps for `GPUBindGroup.insertBinding(resource)`:
+
+ - If `this.[[desc]].layout.[[desc]].dynamicArray` is `undefined`, throw an `OperationError`.
+ - Let `slot` be `this.[[slot_maybe_used]].indexOf(false)`.
+ - If `slot` is `-1`, throw an `OperationError`.
+ - Let `binding` be `slot + this.[[desc]].layout.[[desc]].dynamicArray.start`.
+ - Call `this.update(binding, resource)`.
+ - Return `binding`.
+
+Steps for `GPUBindGroup.removeBinding(binding)`:
+
+ - If `this.[[desc]].layout.[[desc]].dynamicArray` is `undefined`, throw an `OperationError`.
+ - Let `slot` be `binding - this.[[desc]].layout.[[desc]].dynamicArray.start`.
+ - If any of the following is not satisfied, throw an `OperationException`:
+
+    - `slot < this.[[desc]].dynamicArraySize`
+    - `!this.[[slot_maybe_used]][slot]`
+
+ - When `this.[[device]].queue.onSubmittedWorkDone()` is resolved, set `this.[[slot_maybe_used]][slot]` to `false`. (TODO [#5379](https://github.com/gpuweb/gpuweb/issues/5379) actually, make it happen just before all the user-visible `onSubmittedWorkDone`).
```

Comment:
This means the app is required to call onSubmittedWorkDone for this to happen, right?

Reply:
The intent is that it is as-if `onSubmittedWorkDone.then(() => {this.[[slot_maybe_used]][slot] = false;})` was executed, but that this `onSubmittedWorkDone` is resolved before all the user ones for the same queue execution point.

Reply:
Nvm the edits make it clear that yes.